[Compiler Bug] Enum.from_string() produces invalid C output

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.5.1, press to see full `v doctor` output</summary>

|V full version      |V 0.5.1 0c3183c55b39534f9bb0d2f796bb575d39c9d229
|:-------------------|:-------------------
|OS                  |linux, "EndeavourOS Linux"
|Processor           |8 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz
|Memory              |3.87GB/19.29GB
|                    |
|V executable        |/usr/lib/vlang/v
|V last modified time|2026-07-04 08:19:04
|                    |
|V home dir          |NOT writable, value: /usr/lib/vlang
|VMODULES            |OK, value: /home/specter/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/specter
|                    |
|Git version         |git version 2.55.0
|V git status        |N/A
|.git/config present |false
|                    |
|cc version          |cc (GCC) 16.1.1 20260625
|gcc version         |gcc (GCC) 16.1.1 20260625
|clang version       |clang version 22.1.6
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |Error: fatal: detected dubious ownership in repository at '/usr/lib/vlang/thirdparty/tcc'
To add an exception for this directory, call:

	git config --global --add safe.directory /usr/lib/vlang/thirdparty/tcc
 Error: fatal: detected dubious ownership in repository at '/usr/lib/vlang/thirdparty/tcc'
To add an exception for this directory, call:

	git config --global --add safe.directory /usr/lib/vlang/thirdparty/tcc

|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.43

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  test.v && test`
```v
enum Example {
	exam1
	exam2
}

fn main() {
	str := 'example'

	if value := Example.from_string(str) {
		match value {
			.exam1 { println('exam1') }
			.exam2 { println('exam2') }
		}
	}
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/test.01KX03MBWEWNXRKA60TNWQREF1.tmp.c:736: warning: implicit declaration of function 'builtin____new_array_with_default'
cc: /tmp/v_1000/test.01KX03MBWEWNXRKA60TNWQREF1.tmp.c:736: error: '{' expected (got ";")
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

I expected the match to correctly match the enum values to what the string was. (str = exam1, match to .exam1, etc).

**(note)**
Multiple different use cases of from_string() on an enum do not produce valid C, which leads me to believe the issue is from from_string() directly, since bypassing it compiles cleanly.


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -1566,37 +1566,21 @@ fn (mut g Gen) gen_enum_static_from_string(fn_name string, mod_enum_name string,
 	enum_styp := g.styp(enum_typ)
 	option_enum_typ := enum_typ.set_flag(.option)
 	option_enum_styp := g.styp(option_enum_typ)
-	enum_field_names := g.table.get_enum_field_names(mod_enum_name)
-	enum_field_vals := g.table.get_enum_field_vals(mod_enum_name)
+	enum_decl := g.table.enum_decls[mod_enum_name]
+	enum_prefix := g.gen_enum_prefix(enum_typ)
 
 	mut fn_builder := strings.new_builder(512)
 	g.definitions.writeln('${g.static_non_parallel}${option_enum_styp} ${fn_name}(string name);')
 
 	fn_builder.writeln('${g.static_non_parallel}${option_enum_styp} ${fn_name}(string name) {')
 	fn_builder.writeln('\t${option_enum_styp} t1;')
-	fn_builder.writeln('\tbool exists = false;')
-	fn_builder.writeln('\tint inx = 0;')
-	fn_builder.writeln('\tarray field_names = ((array){.data = 0, .offset = 0, .len = 0, .cap = 0, .flags = 0, .element_size = sizeof(string)});')
-	for field_name in enum_field_names {
-		fn_builder.writeln('\tbuiltin__array_push((array*)&field_names, _MOV((string[]){ _S("${field_name}") }));')
+	for field in enum_decl.fields {
+		fn_builder.writeln('\tif (builtin__fast_string_eq(name, _S("${field.name}"))) {')
+		fn_builder.writeln('\t\tbuiltin___option_ok(&(${enum_styp}[]){ ${enum_prefix}${field.name} }, (_option*)&t1, sizeof(${enum_styp}));')
+		fn_builder.writeln('\t\treturn t1;')
+		fn_builder.writeln('\t}')
 	}
-	fn_builder.writeln('\tarray field_vals = ((array){.data = 0, .offset = 0, .len = 0, .cap = 0, .flags = 0, .element_size = sizeof(i64)});')
-	for field_val in enum_field_vals {
-		fn_builder.writeln('\tbuiltin__array_push((array*)&field_vals, _MOV((i64[]){ ${field_val} }));')
-	}
-	fn_builder.writeln('\tfor (${ast.int_type_name} i = 0; i < ${enum_field_names.len}; ++i) {')
-	fn_builder.writeln('\t\tif (builtin__fast_string_eq(name, (*(string*)builtin__array_get(field_names, i)))) {')
-	fn_builder.writeln('\t\t\texists = true;')
-	fn_builder.writeln('\t\t\tinx = i;')
-	fn_builder.writeln('\t\t\tbreak;')
-	fn_builder.writeln('\t\t}')
-	fn_builder.writeln('\t}')
-	fn_builder.writeln('\tif (exists) {')
-	fn_builder.writeln('\t\tbuiltin___option_ok(&(${enum_styp}[]){ (*(i64*)builtin__array_get(field_vals, inx)) }, (_option*)&t1, sizeof(${enum_styp}));')
-	fn_builder.writeln('\t\treturn t1;')
-	fn_builder.writeln('\t} else {')
-	fn_builder.writeln('\t\treturn (${option_enum_styp}){ .state=2, .err=_const_none__, .data={E_STRUCT} };')
-	fn_builder.writeln('\t}')
+	fn_builder.writeln('\treturn (${option_enum_styp}){ .state=2, .err=_const_none__, .data={E_STRUCT} };')
 	fn_builder.writeln('}')
 	g.auto_fn_definitions << fn_builder.str()
 }

--- a/vlib/v/tests/enums/enum_static_from_string_test.v
+++ b/vlib/v/tests/enums/enum_static_from_string_test.v
@@ -12,6 +12,11 @@ enum Color2 as i64 {
 	green
 }
 
+enum Example {
+	exam1
+	exam2
+}
+
 fn test_enum_static_from_string() {
 	color11 := Color1.from_string('red')?
 	println(color11)
@@ -36,4 +41,16 @@ fn test_enum_static_from_string() {
 	color23 := Color2.from_string('bbbbb') or { Color2.unknown }
 	println(color23)
 	assert color23 == Color2.unknown
+}
+
+fn test_enum_static_from_string_if_guard_match() {
+	str := 'exam1'
+	mut matched := ''
+	if value := Example.from_string(str) {
+		match value {
+			.exam1 { matched = 'exam1' }
+			.exam2 { matched = 'exam2' }
+		}
+	}
+	assert matched == 'exam1'
 }

